static/js/blueslip_stacktrace.ts

```diff
@@ -0,0 +1,113 @@
+import * as $ from 'jquery';
+import * as _ from 'underscore';
+import * as StackTraceGPS from 'stacktrace-gps';
+import * as ErrorStackParser from 'error-stack-parser';
+
+type StackFrame = ErrorStackParser.StackFrame;
+
+type CleanStackFrame = {
+    full_path: string;
+    show_path: string;
+    function_name: string;
+    line_number: number;
+};
+
+export function clean_stackframe_data(data: StackFrame): CleanStackFrame {
+    // If the file is local, just show the filename.
+    // Otherwise, show the full path starting from node_modules.
+    const idx = data.fileName.indexOf('node_modules');
+    let full_path;
+    let show_path;
+
+    if (idx !== -1) {
+        full_path = show_path = data.fileName.slice(idx);
+    } else {
+        full_path = data.fileName.split('/').slice(-3).join('/');
+        show_path = full_path.slice(full_path.lastIndexOf('/') + 1);
+    }
+
+    // Don't show the full scope in function name
+    let full_fn_name;
+    try {
+        full_fn_name = /(?:[a-z_]+\.[jt]s(?:\?\w+)?\.)(.+)/.exec(data.functionName)[1];
+    } catch {
+        full_fn_name = data.functionName;
+    }
+    const fn_name = full_fn_name.split('.');
+
+    fn_name[fn_name.length - 1] = '<span class="function-name">'
+        + fn_name[fn_name.length - 1].replace('<', '&lt;').replace('>', '&gt;')
+        + '</span>';
+
+    return {
+        full_path: full_path,
+        show_path: show_path,
+        line_number: data.lineNumber,
+        function_name: fn_name.join('.'),
+    };
+}
+
+const stack_trace_gps = new StackTraceGPS();
+
+async function beautify_stackframe(stackframe: StackFrame): Promise<CleanStackFrame> {
+    return stack_trace_gps
+        .getMappedLocation(stackframe)
+        .then(clean_stackframe_data);
+}
+
+export function display_stacktrace(error: string, stack: string): void {
+    const ex = new Error();
+    ex.stack = stack;
+    const stacktrace: Promise<CleanStackFrame>[] =
+        ErrorStackParser.parse(ex).map(beautify_stackframe);
+
+    // The stacktrace-gps library uses Promise in their APIs for fetching
+    // source maps from the server and the underlying mozilla/source-map
+    // library is compiled to wasm. We assume the developer's browser is recent
+    // enough to support both of these.
```

Comment:
We’re supposed to work on IE11, which doesn’t provide `Promise` natively. But surely we already depend on twelve or fifteen libraries that pull in three different partially-compatible `Promise` polyfills? This is JS we’re talking about here.